zerver/tests/test_users.py

```diff
@@ -460,6 +461,45 @@ def test_access_user_by_id(self) -> None:
             self.example_user("cordelia"), self.example_user("aaron").id, for_admin=False
         )
 
+    def test_access_user_by_id_including_cross_realm(self) -> None:
+        iago = self.example_user("iago")
+        internal_realm = get_realm(settings.SYSTEM_BOT_REALM)
+
+        # Must be a valid user ID in the realm
+        with self.assertRaises(JsonableError):
+            access_user_by_id_including_cross_realm(iago, 1234, for_admin=False)
+        with self.assertRaises(JsonableError):
+            access_user_by_id_including_cross_realm(
+                iago, self.mit_user("sipbtest").id, for_admin=False
+            )
+
+        # Can only access bot users if allow_bots is passed
+        bot = get_system_bot(settings.WELCOME_BOT, internal_realm.id)
+        access_user_by_id_including_cross_realm(iago, bot.id, allow_bots=True, for_admin=False)
+        with self.assertRaises(JsonableError):
+            access_user_by_id_including_cross_realm(iago, bot.id, for_admin=False)
+
+        # Can only access deactivated users if allow_deactivated is passed
+        hamlet = self.example_user("hamlet")
+        do_deactivate_user(hamlet, acting_user=None)
+        with self.assertRaises(JsonableError):
+            access_user_by_id_including_cross_realm(iago, hamlet.id, for_admin=False)
+        with self.assertRaises(JsonableError):
+            access_user_by_id_including_cross_realm(iago, hamlet.id, for_admin=True)
+        access_user_by_id_including_cross_realm(
+            iago, hamlet.id, allow_deactivated=True, for_admin=True
+        )
+
+        # Non-admin user can't admin another user
+        with self.assertRaises(JsonableError):
+            access_user_by_id_including_cross_realm(
+                self.example_user("cordelia"), self.example_user("aaron").id, for_admin=True
+            )
+        # But does have read-only access to it.
+        access_user_by_id_including_cross_realm(
+            self.example_user("cordelia"), self.example_user("aaron").id, for_admin=False
+        )
```

Comment:
So this is a bit subtle, but rather than copy-pasting this test, we should instead just have the existing test verify both cases -- it's much more likely that a lazy developer will avoid forking these two test setups that way.